WordPress/gutenberg
# Try: Navigation Menu Sidebar

Part of https://github.com/WordPress/gutenberg/issues/36667. 

This is a wide exploration PR which adds a new sidebar to the site and post editor to help with navigation menu management.

⚠️ **PLEASE NOTE:** A smaller, more targettted PR is now available in https://github.com/WordPress/gutenberg/pull/39290/ which focuses solely on the Navigation sidebar without addressing wider concerns about the "W" sidebar and global styles.

### After

The new navigation sidebar reuses the ListView component and allows users to easily reorder items even when the menu might be hidden (for example in a mobile menu).

https://user-images.githubusercontent.com/1270189/154540759-00d097a2-4231-4ab0-89ba-224ddc86b76f.mp4

### Before

Previously, clicking on the open list view button would open a modal.

| Open List View | Modal |
|-----|-----|
|  <img width="531" alt="CleanShot 2022-02-10 at 13 44 02@2x" src="https://user-images.githubusercontent.com/1270189/153501519-f57ab90b-63fa-42dc-bcb7-a503f3625d9b.png"> | <img width="668" alt="CleanShot 2022-02-10 at 13 43 00@2x" src="https://user-images.githubusercontent.com/1270189/153501412-92ce268b-572a-4f92-bb60-258f3efe1104.png"> |


### Testing Instructions

- Verify that in the Site Editor, opening list view now opens the new sidebar as expected
- We should see a loader when waiting for menu results
- Verify that in the Widget, Navigation and Post editor that the navigation sidebar is not available.

### Notes

Navigation blocks can store their data in [many ways](https://github.com/WordPress/gutenberg/blob/ca3c6dab4b0fbed1fb67d958a1ac1dd81609eb96/packages/block-library/src/navigation/use-navigation-entities.js):

1. With a reference to a wp_navigation post id
2. `__unstableLocation` such as a "primary" menu. The site editor incorrectly renders an empty placeholder in this case. Note that I won't address that in this PR.
3. With data set in innerBlocks.
4. A navigation block may also be populated from a "classic" menu

#### Wiring up the list view button on the Navigation block toolbar is a package puzzle.

- Each sidebar is namespaced to a particular editor instance, in case functionality needs to diverge. So for example the action we dispatch to open the sidebar in site editor is: 
```jsx
enableComplementaryArea(
			'core/edit-site',
			'edit-site/navigation-menu'
		); 
```
while in the post editor it is: 
```jsx
enableComplementaryArea(
			'core/edit-post',
			'edit-post/navigation-menu'
		);
```
- block-library should not know about a particular editor instance
- the toolbar in block-editor should not know about particular blocks

To solve this, I opted to add an unstable slot/fill following the pattern set by `__unstableBlockSettingsMenuFirstItem`. If folks have a better idea of how to untangle this, I'm happy to hear about suggestions.
